### ISSUE TITLE
[RMP] Unify and clean up block-API in TensorFlow & PyTorch

### ISSUE THREAD

**Issue by marcromeyn** [NVIDIA-Merlin/Merlin#893]
## Problem:
We are currently in a situation where some customers are using merlin-models & some T4Rec to train models. The APIs of these 2 tools have diverged quite dramatically and some features (like extracting embeddings out of models) are only supported in Merlin Models. Both tools require some work in order to have easy to use APIs.

On the Merlin models side, we are in a in-between state where (because of time pressure) there are a bunch of V1 & V2 classes. We would like to migrate all our users to the V2 classes (while removing V2 from the name) & deprecate the old classes.

On the T4Rec side, we would like to keep using this project for session-based models in PyTorch because of the traction we've got. The idea would be to break out the core model-building parts (block-API) in favor of the pytorch-backend of Merlin Models. This roadmap-level ticket focusses on this new pytorch-backend, integration into T4Rec is left out for later. The first major deliverable of this backend is the creation of retrieval models, this because we typically frame session-based models as retrieval-models

## Goal:
Reach feature parity & rough API parity between TF & PyTorch backends in Merlin models.

## New Functionality
- Models
  - PyTorch: New backend, build from the ground up based on the TF implementation. Port the all retrieval examples.
  - TF: Migration to V2 classes, this includes input-blocks, output-blocks and models.

## Constraints:
- We focus on just _retrieval-models_. Ranking-models will be tackled in a future roadmap ticket. 
- Migrating T4Rec to the new Block-API is future work and will be captured in another roadmap-level ticket.

## Starting Point:
In order to properly plan out the work, a [dev-branch](https://github.com/NVIDIA-Merlin/models/tree/torch/dev) is created to answer various design-questions around being able to create retrieval-models in PyTorch. This has lead to a rough MVP that contains all the major pieces. This has also given us a better idea how to break things down to turn the MVP into a fully fleshed product. 

There are 2 design-questions that still need answering:
1. How can we leverage [torchrec](https://github.com/pytorch/torchrec)?
2. How should we represent ragged-tensors in PyTorch? _This question have a dependency on the first one since it seems like torchrec comes with a ragged-tensor representation._

We are planning to have people work in parallel on 4 different major parts: _inputs, outputs, models & masking_.

### Implement base-classes of block-API in PyTorch
_People: @marcromeyn_ 

Currently the block-API is T4Rec is using a similar design to Keras to allow for modules that lazily initialize their variables. We would like to deprecate this in favor of a native way to achieve the same thing that could launched recently. 

- [ ] Implement Block, TabularBlock & ParallelBlock
- [ ] Implement SequentialBlock

### Masking
_People: @sararb, @gabrielspmoreira & @marcromeyn_

This work is dependent on answering the design-question how to handle ragged-tensors.

#### TF
Tasks: _TODO_

#### PyTorch
Tasks: _TODO_

### Input-blocks
_People: @oliverholworthy & @marcromeyn_

#### TF
- [ ] Create migration plan how to move to `InputBlockV2`

#### PyTorch
_Starting point: [MVP](https://github.com/NVIDIA-Merlin/models/tree/torch/dev/merlin/models/torch/inputs)_

- [ ] Implement `Continuous` & `Embeddings`
- [ ] Implement `TabularInputBlock`
- [ ] Implement `Encoder`
- [ ] Add support for sequential-features in input-blocks
- [ ] Do performance testing of holding multiple features in a single embedding-table

### Output-blocks
_People: @gabrielspmoreira & @marcromeyn_

#### TF
- [ ] Create migration plan how to move to `ModelOutput`s & deprecate `PredictionTask`s

#### PyTorch
_Starting point: [MVP](https://github.com/NVIDIA-Merlin/models/tree/torch/dev/merlin/models/torch/outputs)_

- [ ] Port ModelOutput
- [ ] Port RegressionOutput & BinaryOutput
- [ ] Port CategoricalOutput
- [ ] Port ContrastiveOutput + negative samplers
- [ ] Port TopKOutput
- [ ] Port OutputBlock (for multi-task learning)

### Models
_People: @edknv & @marcromeyn_

#### TF
- [ ] Create migration plan how to move to V2 model-classes

#### PyTorch
_Starting point: [MVP](https://github.com/NVIDIA-Merlin/models/tree/torch/dev/merlin/models/torch/models)_

One of the leading questions in the initial experimentation phase was to figure out if we can leverage PyTorch lightning for a high-level training-API (similar to how we use Keras on the TF-side). We are confident that PyTorch Lightning is the right path forward.

- [ ] Implement Model class (using PyTorch lightning)
- [ ] Create custom Trainer that can handle multi-GPU with data-loader
- [ ] Implement RetrievalModel class
- [ ] Port `MatrixFactorizationModel`, `TwoTowerModel` & `YoutubeDNNRetrievalModel`

**Comment by viswa-nvidia** [comment]
@marcromeyn , please create the tasks for PyT and create the tickets so that we can assign them

**Comment by EvenOldridge** [comment]
@marcromeyn @gabrielspmoreira can you work to split this up into: Ranking, Retrieval and Session based

**Comment by marcromeyn** [comment]
Ranking ticket is here: https://github.com/NVIDIA-Merlin/Merlin/issues/1044
